src/oidc_child/oidc_child.c

```diff
@@ -0,0 +1,1425 @@
+                                                         dc_ctx->device_code);
+    if (post_data == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to generate POST data.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    if (client_secret != NULL) {
+        post_data = talloc_asprintf_append(post_data, "&client_secret=%s",
+                                           client_secret);
+        if (post_data == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "Failed to add client secret to POST data.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+    }
+
+    curl_ctx = curl_easy_init();
+    if (curl_ctx == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to initialize curl.\n");
+        ret = EIO;
+        goto done;
+    }
+
+    ret = set_http_opts(curl_ctx, dc_ctx, dc_ctx->token_endpoint, post_data,
+                        NULL, headers);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to set http options.\n");
+        goto done;
+    }
+
+    do {
+        clean_http_data(dc_ctx);
+
+        res = curl_easy_perform(curl_ctx);
+        if (res != CURLE_OK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to send token request.\n");
+            ret = EIO;
+            goto done;
+        }
+
+        ret = parse_token_result(dc_ctx);
+        if (ret != EAGAIN) {
+            break;
+        }
+
+        /* only run once after getting the device code to tell the IdP we are
+         * expecting that the user will connect */
+        if (get_device_code) {
+            /* TODO: check code */
+            ret = EOK;
+            break;
+        }
+
+        sleep(dc_ctx->interval);
+        waiting_time += dc_ctx->interval;
+    } while (waiting_time < dc_ctx->expires_in);
+
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to get token.\n");
+    }
+
+done:
+
+    curl_slist_free_all(headers);
+    curl_easy_cleanup(curl_ctx);
+    return ret;
+}
+
+static errno_t get_userinfo(struct devicecode_ctx *dc_ctx)
+{
+    int ret;
+
+    clean_http_data(dc_ctx);
+    ret = do_http_request(dc_ctx, dc_ctx->userinfo_endpoint, NULL,
+                          dc_ctx->access_token_str);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to send userinfo request.\n");
+    }
+
+    return ret;
+}
+
+#define IN_BUF_SIZE 4096
+static errno_t read_device_code_from_stdin(struct devicecode_ctx *dc_ctx)
+{
+    uint8_t buf[IN_BUF_SIZE];
+    ssize_t len;
+    errno_t ret;
+    char *str;
+
+    errno = 0;
+    len = sss_atomic_read_s(STDIN_FILENO, buf, IN_BUF_SIZE);
+    if (len == -1) {
+        ret = errno;
+        ret = (ret == 0) ? EINVAL: ret;
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "read failed [%d][%s].\n", ret, strerror(ret));
+        return ret;
+    }
+
+    if (len == 0 || *buf == '\0') {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing device code\n");
+        return EINVAL;
+    }
+
+    str = talloc_strndup(dc_ctx, (char *) buf, len);
```

Comment:
memset buf to zero?

Reply:
As Alexey already mentioned below I new use a destructor and `sss_erase_talloc_mem_securely`.